s02/2.md

```diff
@@ -10,14 +10,556 @@ Tartalom
+            return *this; // visszaadjuk a referenciát
+        }
+};
+```
+
+#### Paraméterátadás
+
+Ahogy a változókat, úgy a paramétereket is háromféleképpen tudjuk átadni: `érték szerint`, `referenciaként`, `pointerként`
+
+Az érték szerinti átadás sokszor költséges lehet, mert ekkor a paraméterek másolódnak. A pointer és a referencia közötti
+döntést pedig az határozza meg, hogy míg a pointerek felvehetik a `NULL` értéket, addig a referenciák nem.
+
+Tehát a következők szerint érdemes a paraméterátadást használni:
+
+* _érték szerinti_:
+    + a függvény **nem módosítja a paramétert**
+    + **könnyű másolni**
+    
+    (egyszerű típusok esetén ajánlott, mint pl.: `int`, `double`, `char`, `bool`, stb. Az `std::string` és
+    a különböző STL konténerek nem ajánlottak)
+* _konstans pointer szerinti_:
+    + a függvény **nem módosítja a paramétert**,
+    + **költséges a másolás**, 
+    + a **NULL valid** érték
+* _pointer szerinti_:
+    + a függvény **módosítja a paramétert**,
+    + **költséges a másolás**,
+    + a **NULL valid** érték
+* _konstans referencia szerinti_:
+    + a függvény **nem módosítja a paramétert**,
+    + **költséges a másolás**
+    * a **NULL nem valid** érték
+* _referencia szerinti_: 
+    + a függvény **módosítja a paramétert**
+    + **költséges a másolás**,
+    + a **NULL nem valid** érték
+
+### Move-szemantika
+
+A C++-ban érték szemantika van. Ez egy tiszta memóriaterület szeparációt tud eredményezni, de sokszor
+teljesítmény romlást okozhat nagy objektumok másolása esetén.
+
+Tekintsük a következő `Array` implementációt:
+
+```C++
+class Array{
+    public:
+        Array (const Array&);
+        Array& operator=(const Array&);
+        ~ Array ();
+    private:
+        double *val;
+};
+Array operator+(const Array& left,const Array& right){
+    Array res = left;
+    res += right;
+    return res;
+}
+```
+
+Az `Array` egy osztály, melynek + operátora összekonkatenálja a két paramétert és visszaad egy új listát.
+
+A következő függvény meghívásánál azonban több köztes `Array` példány keletkezik és szűnik meg:
+
+```C++
+void f()
+{
+ Array b, c, d;
+ …
+ Array a = b + c + d;
+}
+```
+
+A move-szemantika az ehhez hasonló problémákra ad megoldást.
+
+* másolás helyett "ellopja" az erőforrást
+* törölhető állapotban hagyja a másik objektumot
+
+Ehhez kell:
+
+* overloadolni lehessen a _Copy-konstruktort_ és az _értékadó operátort_, illetve egyéb függvényeket
+    + meg kell tartani a backward compatibility-t
+    + meg kell különböztetni a bal- és jobbértékeket
+
+#### RValue, LValue
+
+Korábbi nyelvekben értékadás: `<variable> = <expression>` (pl.: `x = 5`)
+
+C/C++-ban értékadás: `<expression> = <expression>` (pl.: `*++ptr = *++qtr`)
+
+* de nem minden esetben működik, pl.: `a+5 = x` helytelen.
+
+**LValue**: kifejezés, mely egy memóriaterületre hivatkozik és lehetővé teszi, hogy a `&` operátorral
+megszerezzük annak memóriacímét
+
+**RValue**: olyan kifejezés, mely nem LValue.
```

Comment:
Esetleg példával? Hogy LValue a `*++ptr` és a `*++qtr` is viszont az `a+5` az RValue?
De ebben nem vagyok biztos, elég homály nekem ez a rész. Ha `a+5` RValue, akkor ezek szerint ezt lehetne move-olni, ugye?